src/marqo/s2_inference/processing/image.py

```diff
@@ -206,20 +208,21 @@ def _load_and_cache_model(self):
         model_type = (self.model_name, self.device)
         model_cache_key = _create_model_cache_key(self.model_name, self.device)
 
-        if model_cache_key not in get_available_models():
-            logger.info(f"loading model {model_type}")
-            if model_type[0] in self.allowed_model_types:
-                func = self.model_load_function
-            else:
-                raise TypeError(f"wrong model for {model_type}")
+        with load_model_lock:
+            if model_cache_key not in get_available_models():
```

Comment:
Is there a reason to acquire the lock even if the model is cached and doesn't need to be loaded?

Reply:
I think acquiring the lock and checking is the correct strategy. This is the scenario:
1. Thread 1 tries to load model B, with model A already in the cache.
2. Thread 2 starts and finds model A in the cache (not in lock)
3. Thread 1 loads model B and ejects model A due to a memory issue
4. Thread 2 tries to use model A to encode and can't find model A (internal error)

BTW, my lock in s2_inference is wrong because I put the existence check outside the lock. I will revise it.